### PR TITLE
Hypothesis testing and generator improving

### PR DESCRIPTION
**Pull request by ArtemiyGranat**

### CODE REVIEW

**Comment by ArtemiyGranat** [comment]
Added clang checks on ubuntu and valgrind found some troubles in code. Will fix them tomorrow. 

MacOS check failed because of `crypt` port. Will try to fix it tomorrow too.

File: main.c [hunk 1]
```diff
@@ -64,6 +64,9 @@ parse_params (config_t *config, int argc, char *argv[])
         case 'r':
           config->brute_mode = BM_RECU;
           break;
+    	case 'y':
+	      config->brute_mode = BM_REC_GEN;
```
Comment (hnarkaytis) [inline main.c:68]: extra indent

File: rec.c [hunk 2]
```diff
@@ -1,5 +1,61 @@
...
+  makecontext (&state->contexts[1], (void (*) (void))brute_rec_gen_helper, 2,
+               config, state);
+
+  swapcontext (&state->contexts[0], &state->contexts[1]);
+}
+
+bool
+rec_state_next (rec_state_t *state)
+{
+  swapcontext (&state->contexts[0], &state->contexts[1]);
```
Comment (hnarkaytis) [inline rec.c:40]: canceled also needs to be checked before context swap. No need to to iterate if iterator is finished.

**Review by hnarkaytis — CHANGES_REQUESTED** [review]

**Comment by ArtemiyGranat** [comment]
Fixed MacOS bruteforce by porting `crypt` lib from musl libc. Now only performance tests on each OS (and valgrind test on Ubuntu with clang) are failing.

File: rec.c [hunk 3]
```diff
@@ -1,5 +1,64 @@
...
+               config, state);
+
+  swapcontext (&state->contexts[0], &state->contexts[1]);
+}
+
+bool
+rec_state_next (rec_state_t *state)
+{
+  if (!state->cancelled)
+    swapcontext (&state->contexts[0], &state->contexts[1]);
```
Comment (hnarkaytis) [inline rec.c:42]: swapcontext might return an error.

File: rec.c [hunk 4]
```diff
@@ -1,5 +1,64 @@
...
+
+  getcontext (&state->contexts[0]);
+  state->contexts[1] = state->contexts[0];
+  state->contexts[1].uc_stack.ss_sp = &state->stack;
+  state->contexts[1].uc_stack.ss_size = sizeof (state->stack);
+  state->contexts[1].uc_link = &state->contexts[0];
+  makecontext (&state->contexts[1], (void (*) (void))brute_rec_gen_helper, 2,
+               config, state);
+
+  swapcontext (&state->contexts[0], &state->contexts[1]);
```
Comment (hnarkaytis) [inline rec.c:35]: swapcontext might return an error.

File: rec.c [hunk 5]
```diff
@@ -1,5 +1,64 @@
...
+  state->cancelled = true;
+}
+
+void
+rec_state_init (rec_state_t *state, task_t *task, config_t *config)
+{
+  state->task = task;
+  state->cancelled = false;
+
+  getcontext (&state->contexts[0]);
```
Comment (hnarkaytis) [inline rec.c:27]: getcontext might return an error.

File: rec.c [hunk 6]
```diff
@@ -1,5 +1,64 @@
 #include "rec.h"
 
+#ifndef __APPLE__
+bool
+brute_rec_gen_handler (task_t *task, void *context)
+{
+  (void)task; // to suppress "unused parameter" warning
+  rec_state_t *state = (rec_state_t *)context;
+  swapcontext (&state->contexts[1], &state->contexts[0]);
```
Comment (hnarkaytis) [inline rec.c:9]: swapcontext might return an error.

**Review by hnarkaytis — CHANGES_REQUESTED** [review]

**Review by hnarkaytis — CHANGES_REQUESTED** [review]
Try -gdwarf-4 instead of -ggdb3

**Review by hnarkaytis — APPROVED** [review]
LGTM

**Comment by ArtemiyGranat** [comment]
> Try -gdwarf-4 instead of -ggdb3

It worked, thank you! Performance tests sometimes fail in macOS, but I use my `crypt` port and the semaphore implementation in Linux too, maybe it's something with `macos-latest` in GH Actions, haven't figured it out yet. Should I run performance tests in Github Actions?

**Comment by ArtemiyGranat** [comment]
> > Try -gdwarf-4 instead of -ggdb3
> 
> It worked, thank you! Performance tests sometimes fail in macOS, but I use my `crypt` port and the semaphore implementation in Linux too, maybe it's something with `macos-latest` in GH Actions, haven't figured it out yet. Should I run performance tests in Github Actions?

I mean Github Actions job in `macos-latest`, on my laptop everything works fine.

**Comment by hnarkaytis** [comment]
Macos on github is flaky with performance tests. I skip performance tests on all virtual platforms.
https://github.com/alexanderchuranov/Metaresc/blob/3c4e281f7d29d8bba4ace4fff9bfecdf5c604840/.github/workflows/test.yml#L43

**Comment by ArtemiyGranat** [comment]
> Macos on github is flaky with performance tests. I skip performance tests on all virtual platforms. https://github.com/alexanderchuranov/Metaresc/blob/3c4e281f7d29d8bba4ace4fff9bfecdf5c604840/.github/workflows/test.yml#L43

Ok, added `WITH_PERF_TEST` variable to Makefile.

**Pull request merged**